### PR TITLE
support alluxioruntime with no secret and generate idempotent mount scripts

### LINKED ISSUE #2771: [FEATURES] Support for generating idempotent Alluxio mount commands as a secret or config in Fluid
**What feature you'd like to add:**

Currently, when deploying a dataset based on Alluxio using Fluid, the mount command is run during the setup process and will not be executed again if the Alluxio master node is scheduled to a different node. To address this issue, we propose adding support for generating idempotent Alluxio mount commands as a secret or config in Fluid, mounting it in the Alluxio master pod, and running it as a post-start script.

This would allow users to store the Alluxio mount commands securely and enable them to be executed even if the master node is rescheduled. Additionally, this would simplify the process of setting up Alluxio-based datasets in Fluid.

It's important that the generated script is idempotent to avoid issues with data consistency and performance. This means that the script should be designed to be executed multiple times without changing the end result after the first execution.

**Why is this feature needed:**

### PR DESCRIPTION
**Pull request by xliuqq**
### Ⅰ. Describe what this PR does
1. support alluxio runtime controller rbac with no secret resource.
2. support alluxio runtime controller to generate idempotent mount scripts instead of directly executing multiple mount command.

The above abilities are supported with **single configuration**.

### Ⅱ. Does this pull request fix one issue?

fixes #2771

### Ⅳ. Describe how to verify it
1. existing alluxio prow test `alluxio_dynamic_mountpoint.py`.
2. manul test for secret logic.

### Ⅴ. Special notes for reviews

### CODE REVIEW

File: charts/alluxio/templates/master/statefulset.yaml [hunk 1]
```diff
@@ -173,11 +173,25 @@ spec:
           {{- if .Values.master.resources  }}
 {{ include "alluxio.master.resources" . | indent 10 }}
           {{- end }}
+          {{- if eq .Values.master.mountConfigStorage "configmap"}}
+          command: ["/etc/fluid/scripts/entrypoint_replace_secrets.sh"]
```
Comment (cheyang) [inline charts/alluxio/templates/master/statefulset.yaml:177]: I think we also name the script `/etc/fluid/scripts/entrypoint.sh` .

File: pkg/ddc/alluxio/transform_volumes.go [hunk 2]
```diff
@@ -54,6 +57,38 @@ func (e *AlluxioEngine) transformMasterVolumes(runtime *datav1alpha1.AlluxioRunt
...
 }
 
+func (e *AlluxioEngine) transformEncryptOptionsToMasterVolumes(dataset *datav1alpha1.Dataset, value *Alluxio) (options map[string]string) {
+	for _, m := range dataset.Spec.Mounts {
+		if common.IsFluidNativeScheme(m.MountPoint) {
+			continue
+		}
+		options = make(map[string]string)
+		for _, encryptOpt := range append(dataset.Spec.SharedEncryptOptions, m.EncryptOptions...) {
+			secretName := encryptOpt.ValueFrom.SecretKeyRef.Name
```
Comment (cheyang) [inline pkg/ddc/alluxio/transform_volumes.go:67]: I am wondering if  a new key is added,  there will be a new file for this key?

File: charts/alluxio/templates/master/statefulset.yaml [hunk 3]
```diff
@@ -359,6 +377,21 @@ spec:
           {{- end }}
       restartPolicy: Always
       volumes:
+        {{- if eq .Values.master.mountConfigStorage "configmap"}}
+        - name: alluxio-mount-conf
+          configMap:
+            name: {{ template "alluxio.fullname" . }}-mount-config
+            items:
+              - key: entrypoint.replace.secrets.script
+                path: entrypoint_replace_secrets.sh
```
Comment (cheyang) [inline charts/alluxio/templates/master/statefulset.yaml:386]: How about changing mode from 555 -> 550?

File: pkg/ddc/alluxio/transform_volumes.go (hunk 2 above)
Comment (xliuqq) [reply pkg/ddc/alluxio/transform_volumes.go:67]: Yes, see https://kubernetes.io/docs/concepts/configuration/secret/#using-secrets-as-files-from-a-pod.
However, should we check the secret key file exist before executing `alluxio fs mount` in script "mount.sh"?

File: pkg/ddc/alluxio/transform.go [hunk 4]
```diff
@@ -356,6 +357,19 @@ func (e *AlluxioEngine) transformMasters(runtime *datav1alpha1.AlluxioRuntime,
...
+	if IsMountWithConfigMap() {
+		e.Log.Info("use configmap to generate mount info")
+		// transform mount secret options
+		nonNativeMounts, err := e.generateNonNativeMountsInfo(dataset)
+		if err != nil {
+			e.Log.Error(err, "generate non native mount info occurs error")
+			return err
+		}
+		value.Master.NonNativeMounts = nonNativeMounts
+		value.Master.MountConfigStorage = "configmap"
```
Comment (hahchenchen) [inline pkg/ddc/alluxio/transform.go:370]: This value "configmap" has been used in many places,can we  replaced the value "configmap" here by a constant？

File: charts/alluxio/templates/config/alluxio-mount-conf.yaml [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+    export ALLUXIO_JAVA_OPTS=$(echo "$ALLUXIO_JAVA_OPTS" | sed "s|/etc/fluid/secrets\(/[-._a-zA-Z0-9]\+\)\{2\}|\`cat &\`|g" )
+    echo "replace $ALLUXIO_JAVA_OPTS"
+    /entrypoint.sh $@
+  mount.script: |
+    #!/usr/bin/env bash
+    # set the -e option, when some command return non zero, the shell will exit, no need to check the executed status.
+    set -xe
+
+    # check alluxio master ready
+    alluxio fsadmin report
```
Comment (hahchenchen) [inline charts/alluxio/templates/config/alluxio-mount-conf.yaml:50]: Can we check the report result if is ready? if not ready, exit -1 or some bad code. 
I think we can check the result by using  the shell Return Code or Regular matching.

File: pkg/ddc/alluxio/ufs_internal.go [hunk 6]
```diff
@@ -201,32 +241,69 @@ func (e *AlluxioEngine) processUpdatingUFS(ufsToUpdate *utils.UFSToUpdate) (err
...
+	mountedPaths = utils.RemoveString(mountedPaths, "/")
+
+	var datasetMountPaths []string
+	for _, mount := range dataset.Spec.Mounts {
+		if common.IsFluidNativeScheme(mount.MountPoint) {
+			continue
+		}
+		m := utils.UFSPathBuilder{}.GenAlluxioMountPath(mount)
+		// skip root path mount
+		if m != "/" {
```
Comment (hahchenchen) [inline pkg/ddc/alluxio/ufs_internal.go:297]: Some times we also change the root path, should we skip it?

File: charts/alluxio/templates/config/alluxio-mount-conf.yaml (hunk 5 above)
Comment (xliuqq) [reply charts/alluxio/templates/config/alluxio-mount-conf.yaml:50]: `set -e` checks the command exeuting result. If a command returns non zero, the script will exits.

File: pkg/ddc/alluxio/ufs_internal.go (hunk 6 above)
Comment (xliuqq) [reply pkg/ddc/alluxio/ufs_internal.go:297]: In my opinion，Alluxio does not support change the root path.

File: charts/alluxio/templates/master/statefulset.yaml (hunk 1 above)
Comment (xliuqq) [reply charts/alluxio/templates/master/statefulset.yaml:177]: ok, done.

File: charts/alluxio/templates/master/statefulset.yaml (hunk 3 above)
Comment (xliuqq) [reply charts/alluxio/templates/master/statefulset.yaml:386]: ok, done.

File: pkg/ddc/alluxio/transform.go (hunk 4 above)
Comment (xliuqq) [reply pkg/ddc/alluxio/transform.go]: ok, done.

File: pkg/ddc/alluxio/ufs.go [hunk 7]
```diff
@@ -18,12 +18,20 @@ package alluxio
...
 	"fmt"
-
 	datav1alpha1 "github.com/fluid-cloudnative/fluid/api/v1alpha1"
 	"github.com/fluid-cloudnative/fluid/pkg/utils"
 	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"os"
 )
 
+func IsMountWithConfigMap() bool {
+	if envVal, exists := os.LookupEnv(MountConfigStorage); exists {
```
Comment (wangshli) [inline pkg/ddc/alluxio/ufs.go:28]: For runtime controller, the env MountConfigStorage is always configmap, and it does not change with different runtimes. So it is need more information to judge if the runtime IsMountWithConfigMap.

File: pkg/ddc/alluxio/ufs.go [hunk 8]
```diff
@@ -18,12 +18,20 @@ package alluxio
...
+		return envVal == ConfigmapStorageName
+	}
+	// default value
+	return true
```
Comment (wangshli) [inline pkg/ddc/alluxio/ufs.go:32]: For old dataset (using secret), it should not return true here.

File: pkg/ddc/alluxio/ufs_internal.go [hunk 9]
```diff
@@ -201,32 +241,69 @@ func (e *AlluxioEngine) processUpdatingUFS(ufsToUpdate *utils.UFSToUpdate) (err
...
-		err = e.Client.Status().Update(context.TODO(), datasetToUpdate)
+
+	return true, nil
+}
+
+// update alluxio mount using script in configmap
+func (e *AlluxioEngine) updateUFSWithMountConfigMapScript(dataset *datav1alpha1.Dataset) (updateReady bool, err error) {
+	// 1. update non native mount info according the data.Spec.Mounts
+	mountConfigMapName := e.getMountConfigmapName()
+	mountConfigMap, err := kubeclient.GetConfigmapByName(e.Client, mountConfigMapName, e.namespace)
```
[the remark this answers is not in the document]
Comment (xliuqq) [reply pkg/ddc/alluxio/ufs_internal.go:257]: Thanks, fixed.
When configmap not found, condsidered as old dataset (using secret), record event and log not support error.

File: pkg/ddc/alluxio/transform_test.go [hunk 10]
```diff
@@ -1033,3 +1033,163 @@ func TestTransformFuseProperties(t *testing.T) {
...
+									SecretKeyRef: datav1alpha1.SecretKeySelector{
+										Name: "ds-secret",
+										Key:  "secret-key",
+									},
+								},
+							},
+						},
+					}, {
+						MountPoint: "local:///mnt/test",
+						Name:       "local",
```
Comment (cheyang) [inline pkg/ddc/alluxio/transform_test.go:1076]: Please define a constant instead of duplicating this literal "local:///mnt/test" 3 times.

Comment (xliuqq) [reply pkg/ddc/alluxio/transform_test.go]: fixed.

File: charts/alluxio/templates/master/statefulset.yaml [hunk 11]
```diff
@@ -173,11 +173,25 @@ spec:
...
+          command: ["/etc/fluid/scripts/entrypoint.sh"]
+          {{ else }}
           command: ["/entrypoint.sh"]
+          {{- end }}
           {{- if .Values.master.args }}
           args:
 {{ toYaml .Values.master.args | trim | indent 12 }}
           {{- end }}
+          {{- if eq .Values.master.mountConfigStorage "configmap"}}
+          startupProbe:
```
Comment (cheyang) [inline charts/alluxio/templates/master/statefulset.yaml:186]: If the version of K8s is less than 1.18, can we use postStart to replace?

File: charts/alluxio/templates/master/statefulset.yaml [hunk 12]
```diff
@@ -173,11 +173,25 @@ spec:
...
+            exec:
+              command:
+                - "/etc/fluid/scripts/mount.sh"
```
Comment (cheyang) [inline charts/alluxio/templates/master/statefulset.yaml:189]: I suggest adding `>> /proc/1/fd/1` to help the user know what happened when debgging.

File: charts/fluid/fluid/values.yaml [hunk 13]
```diff
@@ -61,6 +61,7 @@ runtime:
       image: alluxio/alluxio-dev:2.9.0
     fuse:
       image: alluxio/alluxio-dev:2.9.0
+    mountConfigStorage: configmap
```
[the remark this answers is not in the document]
Comment (xliuqq) [reply charts/fluid/fluid/values.yaml:64]: fluid old version does not store the mount info into a storage. If the value is not 'configmap', the mount logic and extracting secrets is the same as before. Currently no plan for support other storage.

File: charts/alluxio/templates/config/alluxio-mount-conf.yaml [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+    mounted_dst=$(echo "$mounted" | cut -d " " -f 3 | { grep -x -v "/" || true; } )
+
+    # mount.info content like below:
+    # /spark https://mirrors.bit.edu.cn/apache/spark --option k=v --readonly
+    # get all needed mount path, ignore the root path '/', "grep -x" match the line
+    need_mount_dst=$(cat ${base_path}/mount.info | cut -d " " -f 1 | { grep -x -v "/" || true; })
+
+    # unmount, find already mounted in alluxio but not existed in mount.info
+    # -x match the line, -v find the unmatched, -F pattern
+    unmount_dst=$(echo "$mounted_dst" | { grep -x -v -F "$need_mount_dst" || true; })
```
Comment (TrafalgarZZZ) [inline charts/alluxio/templates/config/alluxio-mount-conf.yaml:71]: If there are multiple line in `mount.info` what will happen when we `grep -x -v -F "$need_mount_dst"`?

Comment (TrafalgarZZZ) [reply charts/alluxio/templates/config/alluxio-mount-conf.yaml:71]: Plus, shall we also check if the mount src changed?

Comment (xliuqq) [reply charts/alluxio/templates/config/alluxio-mount-conf.yaml:71]: not match any of line .

Comment (xliuqq) [reply charts/alluxio/templates/config/alluxio-mount-conf.yaml:71]: > Plus, shall we also check if the mount src changed?

I think it is a new feature,  as the  [AnalyzePathsDelta](https://github.com/fluid-cloudnative/fluid/blob/1d15d54cbabf61dd7f234212d5f8da14ebe35c9e/pkg/utils/dataset.go#L152) function only use alluxio mount path (not src path) to recgonize dataset mount differences.

Comment (TrafalgarZZZ) [reply charts/alluxio/templates/config/alluxio-mount-conf.yaml:71]: > > Plus, shall we also check if the mount src changed?
> 
> I think it is a new feature, as the [AnalyzePathsDelta](https://github.com/fluid-cloudnative/fluid/blob/1d15d54cbabf61dd7f234212d5f8da14ebe35c9e/pkg/utils/dataset.go#L152) function only use alluxio mount path (not src path) to recgonize dataset mount differences.

Thx. Now I think this is okay.

File: charts/alluxio/templates/config/alluxio-mount-conf.yaml [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+  {{- end }}
+data:
+  mount.info: |
+    {{- range $index, $value :=.Values.master.nonNativeMounts }}
+    {{$value}}
+    {{- end}}
+  entrypoint.script: |
+    #!/usr/bin/env bash
+    export ALLUXIO_JAVA_OPTS=$(echo "$ALLUXIO_JAVA_OPTS" | sed "s|/etc/fluid/secrets\(/[-._a-zA-Z0-9]\+\)\{2\}|\`cat &\`|g" )
+    echo "replace $ALLUXIO_JAVA_OPTS"
```
[the remark this answers is not in the document]
Comment (xliuqq) [reply charts/alluxio/templates/config/alluxio-mount-conf.yaml:42]: ok,fixed.

File: charts/alluxio/templates/master/statefulset.yaml (hunk 12 above)
Comment (xliuqq) [reply charts/alluxio/templates/master/statefulset.yaml:189]: ok, fixed.

File: charts/alluxio/templates/master/statefulset.yaml [hunk 16]
```diff
@@ -173,11 +173,25 @@ spec:
...
+            initialDelaySeconds: 5
+            failureThreshold: 3
+            timeoutSeconds: 15
+            periodSeconds: 20
```
[the remark this answers is not in the document]
Comment (xliuqq) [reply charts/alluxio/templates/master/statefulset.yaml:192]: In my environment, the startup probe will  be success at the second time. Increase the failureThreshold to 5 or 10 ?

[the remark this answers is not in the document]
Comment (cheyang) [reply charts/alluxio/templates/master/statefulset.yaml:192]: I suggest using 10 by default.

[the remark this answers is not in the document]
Comment (xliuqq) [reply charts/alluxio/templates/master/statefulset.yaml]: ok, fixed

**Review by cheyang — APPROVED** [review]
/approve
/lgtm

**Comment by cheyang** [comment]
Thanks for your great contribution! @xliuqq

**Pull request merged**